File: src/main/ruby/g0401_0500/s0494_target_sum/solution.rb
# #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Backtracking
# #Big_O_Time_O(n*(sum+s))_Space_O(n*(sum+s))
# #2023_11_27_Time_149_ms_(76.92%)_Space_211.8_MB_(100.00%)

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def find_target_sum_ways(nums, s)
  sum = nums.sum
  s = s.abs

  # Invalid s, just return 0
  return 0 if s > sum || (sum + s).odd?

  dp = Array.new((sum + s) / 2 + 1) {Array.new(nums.length + 1, 0)}
  dp[0][0] = 1

  # empty knapsack must be processed specially
  nums.each_with_index do |num, i|
    dp[0][i + 1] = num.zero? ? dp[0][i] * 2 : dp[0][i]
  end

  (1...dp.length).each do |i|
    (0...nums.length).each do |j|
      dp[i][j + 1] += dp[i][j]
      dp[i][j + 1] += dp[i - nums[j]][j] if nums[j] <= i
    end
  end

  dp[(sum + s) / 2][nums.length]
end

File: src/main/python/g0401_0500/s0494_target_sum/Solution.py
# #Medium #Array #Dynamic_Programming #Backtracking #Big_O_Time_O(n*(sum+s))_Space_O(n*(sum+s))
# #2024_06_07_Time_105_ms_(86.96%)_Space_16.7_MB_(80.93%)

class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        total_sum = sum(nums)
        s = abs(target)
        
        # Invalid s, just return 0
        if s > total_sum or (total_sum + s) % 2 != 0:
            return 0
        
        target = (total_sum + s) // 2
        dp = [[0] * (len(nums) + 1) for _ in range(target + 1)]
        dp[0][0] = 1
        
        # empty knapsack must be processed specially
        for i in range(len(nums)):
            if nums[i] == 0:
                dp[0][i + 1] = dp[0][i] * 2
            else:
                dp[0][i + 1] = dp[0][i]
        
        for i in range(1, target + 1):
            for j in range(len(nums)):
                dp[i][j + 1] += dp[i][j]
                if nums[j] <= i:
                    dp[i][j + 1] += dp[i - nums[j]][j]
        
        return dp[target][len(nums)]
